File: algo/lc.160.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        /*
        if (nullptr == headA || nullptr == headB) return nullptr;
        
        // traverse list A, and memorize the nodes
        map<ListNode*, bool> mA;
        ListNode* cur = headA;
        while (cur != nullptr) {
            mA[cur] = true;
        }
        // traverse list B, see if there is any node appeard in list A
        cur = headB;
        map<ListNode*, bool>::iterator mApos;
        while (cur != nullptr) {
            if ((mApos = mA.find(cur)) != mA.end()) {
                return cur;
            }
            cur = cur->next;
        }
        // no intersection at all
        return nullptr; // timeout
        */
        
        if (nullptr == headA || nullptr == headB) return nullptr;
        
        // get len(A) and len(B)
        int lenA = 0, lenB = 0;
        ListNode* curA = headA, * curB = headB;
        while (curA != nullptr) {
            curA = curA -> next;
            lenA++;
        }
        while (curB != nullptr) {
            curB = curB -> next;
            lenB++;
        }
        // the cursor of the longest list go first by (m-n) steps
        curA = headA;
        curB = headB;
        if (lenA != lenB) {
            int s = max(lenA, lenB) - min(lenA, lenB);
            if (lenA > lenB) {
                for (int i = 0; i < s; i++) curA = curA->next;
            } else { // lenA < lenB
                for (int i = 0; i < s; i++) curB = curB->next;
            }
        }
        // move A and B together and see wether they meet
        while (curA != nullptr && curB != nullptr) {
            if (curA == curB) {
                return curA;
            } else {
                curA = curA -> next;
                curB = curB -> next;
            }
        }
        // they didn't meet each other
        return nullptr;
    }
};

File: algo/lc.160.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        # flatten list A
        flatA = []
        cur = headA
        while cur is not None:
            flatA.append(cur)
            cur = cur.next
        # flatten list B
        flatB = []
        cur = headB
        while cur is not None:
            flatB.append(cur)
            cur = cur.next
        # reverse
        flatA = flatA[::-1]
        flatB = flatB[::-1]
        # compare
        intersect = None
        if flatA[0] != flatB[0]:
            return None
        cursor = 0
        while flatA[cursor] == flatB[cursor]:
            intersect = flatA[cursor]
            cursor = cursor + 1
            if cursor >= min(len(flatA), len(flatB)):
                break
        return intersect
        
